### DISCUSSION TITLE
Clustering deterministic deduplication

### DISCUSSION THREAD

**Discussion by JWOsmond** [moj-analytical-services/splink#1173] [Q&A] [answered]
I've done a deterministic link on a dataset of half a million records, using the `linker.deterministic_link()` function. Is there a way of clustering the links, like can be done for a probabilistic linkage (the `linker.cluster_pairwise_predictions_at_threshold()` function)?

**Comment by RobinL** [comment, marked as answer]
You're right, this is an omission.  `cluster_pairwise_predictions_at_threshold` fails at the moment because it's looking for a `match_probability` column.



For the moment, here's a hack that works by creating a `match_probability` column on the outputs of `deterministic_link`, which is set to 1.0 (i.e. a perfect match) for all rows:

But in the longer run, we should explicitly support passing the results of a deterministic link to `cluster_pairwise_predictions_at_threshold`.  I have opened issue #1175 with reference to this discussion

```python 
from splink.duckdb.duckdb_linker import DuckDBLinker
from splink.duckdb.duckdb_comparison_library import exact_match
import pandas as pd

df = pd.read_csv("https://github.com/moj-analytical-services/splink_demos/raw/master/data/fake_1000.csv")

settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.first_name = r.first_name and l.surname = r.surname",
        "l.dob = r.dob and l.surname = r.surname",
    ],
}


linker = DuckDBLinker(df, settings)

df_matches = linker.deterministic_link()
df_matches.as_pandas_dataframe(limit=5)

sql = f"""
select *, 1.00 as match_probability
from {df_matches.physical_name}
"""
deterministic_matches = linker.query_sql(sql, output_type="splink_df")

clusters = linker.cluster_pairwise_predictions_at_threshold(deterministic_matches, 0.5)

clusters.as_pandas_dataframe()
```

**Comment by JWOsmond** [comment]
Thanks Robin that's been really useful. Also, that `query_sql` function more generally is a great help!
